ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/KeyByBigQueryTableDestination.java

```diff
@@ -121,21 +121,23 @@ public TableDestination getTableDestination(Map<String, String> attributes) {
     return tableDestination;
   }
 
-  @Override
-  protected KV<TableDestination, PubsubMessage> processElement(PubsubMessage message) {
-    message = PubsubConstraints.ensureNonNull(message);
-
-    return KV.of(getTableDestination(message.getAttributeMap()), message);
-  }
-
-  @Override
-  public WithErrors.Result<PCollection<KV<TableDestination, PubsubMessage>>> expand(
-      PCollection<PubsubMessage> input) {
-    WithErrors.Result<PCollection<KV<TableDestination, PubsubMessage>>> result = super.expand(
-        input);
-    result.output()
-        .setCoder(KvCoder.of(TableDestinationCoderV3.of(), PubsubMessageWithAttributesCoder.of()));
-    return result;
+  /**
+   * Returns a MapElements transform for retrieving the table destinations.
+   */
+  public MapWithFailures<PubsubMessage, KV<TableDestination, PubsubMessage>, PubsubMessage> map() {
```

Comment:
I don't see how it's possible to refactor this to be a static method such as `LimitPayloadSize.of()` introduced in https://github.com/mozilla/gcp-ingestion/pull/1083/files 
Not sure if there is a better way or name for this in this case?

Reply:
something like this:
```java
  public MapWithFailures<PubsubMessage, KV<TableDestination, PubsubMessage>, PubsubMessage> of(
      ValueProvider<String> tableSpecTemplate, ValueProvider<String> partitioningField,
      ValueProvider<List<String>> clusteringFields) {
    KeyByBigQueryTableDestination transform = new KeyByBigQueryTableDestination(tableSpecTemplate, partitioningField,
        clusteringFields);
    return MapElements.into(TypeDescriptors.kvs(TypeDescriptor.of(TableDestination.class),
        TypeDescriptor.of(PubsubMessage.class))).via((PubsubMessage msg) -> {
          msg = PubsubConstraints.ensureNonNull(msg);
          return KV.of(transform.getTableDestination(msg.getAttributeMap()), msg);
        }).exceptionsInto(TypeDescriptor.of(PubsubMessage.class))
        .exceptionsVia((WithFailures.ExceptionElement<PubsubMessage> ee) -> {
          try {
            throw ee.exception();
          } catch (LimitPayloadSize.PayloadTooLargeException e) {
            return FailureMessage.of(KeyByBigQueryTableDestination.class.getSimpleName(),
                ee.element(), ee.exception());
          }
        });
  }
```

Reply:
Since this class also has the non-static `getTableDestination` method, it might make sense in this case to have `KeyByBigQueryTableDestination` extend `PTransform` and put the `MapElements` call into the `expand` method. That way we can keep the existing `.of` factory method.

Reply:
Yeah, I think that might be the most elegant solution here.